python/raydp/torch/estimator.py

```diff
@@ -156,154 +209,93 @@ def _check(self):
-
-                # create loss
-                if inspect.isclass(outer._loss) and issubclass(outer._loss, TLoss):
-                    loss = outer._loss
-                elif isinstance(outer._loss, TLoss):
-                    loss = outer._loss
-                elif callable(outer._loss):
-                    loss = outer._loss(config)
-                else:
-                    raise Exception(
-                        "Unsupported parameter, we only support torch.nn.modules.loss._Loss "
-                        "subclass, subclass instance or a function(dict -> loss)")
-
-                # create lr scheduler
-                if outer._lr_scheduler_creator:
-                    lr_scheduler = outer._lr_scheduler_creator(optimizer, config)
-                else:
-                    lr_scheduler = None
-
-                registered = self.register(
-                    models=model, optimizers=optimizer, criterion=loss, schedulers=lr_scheduler)
-                if lr_scheduler is not None:
-                    self.model, self.optimizer, self.criterion, self.scheduler = registered
-                else:
-                    self.model, self.optimizer, self.criterion = registered
-
-                # create dataset
-                batch_size = config["batch_size"]
-                get_shard_config = config.get("get_shard", {})
-                if "shuffle" in config:
-                    get_shard_config["shuffle"] = config["shuffle"]
-                if not self._is_distributed:
-                    world_rank = -1
-                else:
-                    world_rank = self.world_rank
-                train_data = train_ds.get_shard(world_rank, **get_shard_config)
-                train_loader = DataLoader(train_data, batch_size=batch_size)
-
-                if evaluate_ds is not None:
-                    evaluate_data = evaluate_ds.get_shard(self.world_rank, **get_shard_config)
-                    evaluate_loader = DataLoader(evaluate_data, batch_size=batch_size)
-                else:
-                    evaluate_loader = None
-
-                self.register_data(train_loader=train_loader, validation_loader=evaluate_loader)
-
-        self._trainer = TorchTrainer(num_workers=self._num_workers,
-                                     training_operator_cls=TorchEstimatorOperator,
-                                     add_dist_sampler=False,
-                                     scheduler_step_freq=self._scheduler_step_freq,
-                                     **self._extra_config)
-
-    def _create_tf_ds(self, ds: MLDataset) -> TorchMLDataset:
-        return ds.to_torch(self._feature_columns,
-                           self._feature_shapes,
-                           self._feature_types,
-                           self._label_column,
-                           self._label_shape,
-                           self._label_type)
+    @staticmethod
+    def train_epoch(dataset, model, criterion, optimizer, scheduler=None):
```

Comment:
Is it necessary to be a static method?

Reply:
I think it could be called by class.func() without creating a class object.

Reply:
I felt a little weird because train_fun is defined out of the class, while these two functions are static methods of the class. Why is it so?

Reply:
I was troubled by this. I want to define train_fun() in the class, but the Trainer would not find it. Maybe I should put train_epoch() and evalue_epoch() together out of the class?

Reply:
Yes, I think that's better, if train_fun cannot be in the class.

Reply:
> I was troubled by this. I want to define train_fun() in the class, but the Trainer would not find it. Maybe I should put train_epoch() and evalue_epoch() together out of the class?

hmmm I retested this, it works.